lib/yargs.ts

```diff
@@ -0,0 +1,1694 @@
+    if (defaultDescription) {
+      assertSingleKey(key)
+      options.defaultDescription[key] = defaultDescription
+    }
+    if (typeof value === 'function') {
+      assertSingleKey(key)
+      if (!options.defaultDescription[key]) options.defaultDescription[key] = usage.functionDescription(value)
+      value = value.call()
+    }
+    populateParserHintSingleValueDictionary<'default'>(self.default, 'default', key, value)
+    return self
+  }
+
+  self.describe = function (key: string | string[] | Dictionary<string>, desc?: string) {
+    argsert('<object|string|array> [string]', [key, desc], arguments.length)
+    setKey(key, true)
+    usage.describe(key, desc)
+    return self
+  }
+
+  function setKey (key: string | string[] | Dictionary<string | boolean>, set?: boolean | string) {
+    populateParserHintSingleValueDictionary(setKey, 'key', key, set)
+    return self
+  }
+
+  function demandOption (keys: string | string[] | Dictionary<string | undefined>, msg?: string) {
+    argsert('<object|string|array> [string]', [keys, msg], arguments.length)
+    populateParserHintSingleValueDictionary(self.demandOption, 'demandedOptions', keys, msg)
+    return self
+  }
+  self.demandOption = demandOption
+
+  self.coerce = function (
+    keys: string | string[] | Dictionary<CoerceCallback>,
+    value?: CoerceCallback
+  ) {
+    argsert('<object|string|array> [function]', [keys, value], arguments.length)
+    populateParserHintSingleValueDictionary(self.coerce, 'coerce', keys, value)
+    return self
+  }
+
+  function populateParserHintSingleValueDictionary<
+    T extends Exclude<DictionaryKeyof<Options>, DictionaryKeyof<Options, any[]>> | 'default',
+    K extends keyof Options[T] & string = keyof Options[T] & string,
+    V extends ValueOf<Options[T]> = ValueOf<Options[T]>
+  > (
+    builder: (key: K, value: V, ...otherArgs: any[]) => YargsInstance,
+    type: T,
+    key: K | K[] | { [key in K]: V },
+    value?: V
+  ) {
+    populateParserHintDictionnary<T, K, V>(builder, type, key, value, (type, key, value) => {
+      options[type][key] = value as ValueOf<Options[T]>
+    })
+  }
+
+  function populateParserHintArrayDictionnary<
+    T extends DictionaryKeyof<Options, any[]>,
+    K extends keyof Options[T] & string = keyof Options[T] & string,
+    V extends ValueOf<ValueOf<Options[T]>> | ValueOf<ValueOf<Options[T]>>[] = ValueOf<ValueOf<Options[T]>> | ValueOf<ValueOf<Options[T]>>[]
+  > (
+    builder: (key: K, value: V, ...otherArgs: any[]) => YargsInstance,
+    type: T,
+    key: K | K[] | { [key in K]: V },
+    value?: V
+  ) {
+    populateParserHintDictionnary<T, K, V>(builder, type, key, value, (type, key, value) => {
+      options[type][key] = (options[type][key] || [] as Options[T][keyof Options[T]]).concat(value)
+    })
+  }
+
+  function populateParserHintDictionnary<T extends keyof Options, K extends keyof Options[T], V> (
+    builder: (key: K, value: V, ...otherArgs: any[]) => YargsInstance,
```

Comment:
You've put an incredible amount of work into building out generics 👏 I think this will make all the difference with regards to yargs' usability.

I think before we release the next major version of yargs, which announces TypeScript, we should add a `.ts` test file, specifically exercising generic behavior. I would be happy to do this, as it will be a good learning experience for me, and help me dig further into your implementation.

Reply:
However, the work on generics is far from being finished, as we still have to integrate the incredible amount of work on parsed arguments generics which had been done in `@types/yargs` 😄